### PR TITLE
Add VertexArrayObject support for WebGL2

### PR DESCRIPTION
**Pull request by imiklos**
The implementation was already in place for OpenGL ES.
My approach with this patch is to add support for WebGL2 by
sharing the implementation between the WebGL2 and GLES.

Sadly I couldn't manage to place the WebGL2 and OpenGL ES variant to the same slot in `WebGLRenderingContext`, therefore I had to do same code duplication.

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] There are tests for these changes

cc @jdm @zakorgy @mmatyas

### CODE REVIEW

File: components/script/dom/webglvertexarrayobjectoes.rs [hunk 1]
```diff
@@ -2,41 +2,28 @@
...
 
 #[dom_struct]
 pub struct WebGLVertexArrayObjectOES {
     webgl_object_: WebGLObject,
-    id: Option<WebGLVertexArrayId>,
-    ever_bound: Cell<bool>,
-    is_deleted: Cell<bool>,
-    vertex_attribs: DomRefCell<Box<[VertexAttribData]>>,
-    element_array_buffer: MutNullableDom<WebGLBuffer>,
+    array_object: Box<VertexArrayObject>,
```
[the remark this answers is not in the document]
Comment (jdm) [reply components/script/dom/webglvertexarrayobjectoes.rs]: Same question for webglvertexarrayobject.rs.

File: components/script/dom/webglrenderingcontext.rs [hunk 2]
```diff
@@ -77,7 +79,7 @@ use pixels::{self, PixelFormat};
 use script_layout_interface::HTMLCanvasDataSource;
 use serde::{Deserialize, Serialize};
 use servo_config::pref;
-use std::cell::Cell;
+use std::cell::{Cell, Ref};
```
Comment (jdm) [inline components/script/dom/webglrenderingcontext.rs]: I suspect we might need to import dom::bindings::cell::Ref instead?

**Review by jdm — CHANGES_REQUESTED** [review]
Looking good!

File: components/script/dom/webglvertexarrayobjectoes.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (imiklos) [reply components/script/dom/webglvertexarrayobjectoes.rs]: No need for the Box here, thank you for pointing that out.

File: components/script/dom/webglrenderingcontext.rs (hunk 2 above)
Comment (imiklos) [reply components/script/dom/webglrenderingcontext.rs]: Yes, you are right! Thank you!

**Comment by imiklos** [comment]
@jdm Thank you for the review! I've updated the pull request with the changes you mentioned.

**Comment by jdm** [comment]
```
  ▶ Unexpected subtest result in /_webgl/conformance2/context/incorrect-context-object-behaviour.html:
  │ FAIL [expected PASS] WebGL test #1: successfullyParsed should be true (of type boolean). Was undefined (of type undefined).
  │   → assert_true: successfullyParsed should be true (of type boolean). Was undefined (of type undefined). expected true got false
  │ 
  │ reportTestResultsToHarness/<@http://web-platform.test:8000/_webgl/js/js-test-pre.js:123:24
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1956:25
  │ test@http://web-platform.test:8000/resources/testharness.js:548:30
  │ reportTestResultsToHarness@http://web-platform.test:8000/_webgl/js/js-test-pre.js:122:15
  │ testFailed@http://web-platform.test:8000/_webgl/js/js-test-pre.js:249:31
  │ shouldBe@http://web-platform.test:8000/_webgl/js/js-test-pre.js:406:19
  │ shouldBeTrue@http://web-platform.test:8000/_webgl/js/js-test-pre.js:432:37
  └ @http://web-platform.test:8000/_webgl/js/js-test-post.js:24:13
```

**Pull request closed without merging**